PyInstaller/building/imphook.py

```diff
@@ -12,18 +12,573 @@
+        if attr_name in _MAGIC_MODULE_HOOK_ATTR_NAMES:
+            self._load_hook_module()
+            return self._hook_attrs[attr_name]
+        # Else, this is an undefined attribute. Raise an exception.
+        else:
+            raise AttributeError(attr_name)
+
+
+    def __setattr__(self, attr_name, attr_value):
+        '''
+        Set the attribute with the passed name to the passed value.
+
+        If this is a magic attribute, this hook script will be lazily loaded
+        before setting this attribute. Unlike `__getattr__()`, this special
+        method is called to set _any_ attribute -- including magic, non-magic,
+        and undefined attributes.
+
+        See Also
+        ----------
+        Class docstring for supported magic attributes.
+        '''
+
+        # If this is a magic attribute, set this attribute to the passed value.
+        if attr_name in _MAGIC_MODULE_HOOK_ATTR_NAMES:
+            # Validate this attribute's new value.
+            if attr_name in {'datas', 'binaries', 'excludedimports'}:
+                if not isinstance(attr_value, Set):
+                    raise AttributeError(
+                        'Magic attribute "{}" unsettable to a "{}". '
+                        'set() expected.'.format(
+                            attr_name, type(attr_value)))
+            elif attr_name in {'hiddenimports',}:
+                if not isinstance(attr_value, Sequence):
+                    raise AttributeError(
+                        'Magic attribute "{}" unsettable to a "{}". '
+                        'list() expected.'.format(
+                            attr_name, type(attr_value)))
+            else:
+                raise AttributeError(
+                    'Magic attribute "{}" unsettable.'.format(attr_name))
+
+            # Initialize this attribute by lazy loading this hook script.
+            self._load_hook_module()
+
+            # Set this attribute.
+            self._hook_attrs[attr_name] = attr_value
+        # Else, this is a non-magic attribute. Defer to our superclass.
+        else:
+            return super(ModuleHook, self).__setattr__(attr_name, attr_value)
+
+
+    def __delattr__(self, attr_name):
+        '''
+        Remove the attribute with the passed name from this object.
+
+        If this is a magic attribute, this hook script will be lazily loaded
+        before removing the attribute. Unlike `__getattr__()`, this special
+        method is called to remove _any_ attribute -- including magic,
+        non-magic, and undefined attributes.
+
+        See Also
+        ----------
+        Class docstring for supported magic attributes.
+        '''
+
+        # If this is a magic attribute, initialize this attribute by lazy
+        # loading this hook script and then remove this attribute.
+        if attr_name in _MAGIC_MODULE_HOOK_ATTR_NAMES:
+            self._load_hook_module()
+            del self.__hook_attrs[attr_name]
```

Comment:
> Hm, this is asymmetric: Line 268 defines default values, but here one can delete the attribute.

But... isn't that the case with most instance variables initialized by an `__init__()` method?

> You could save this `__delattr__` anyway if storing the hook attrs in `self`...

I think I see what you're getting at. _Maybe?_

For each hook attribute (e.g., `self._hook_module.datas`), you propose dynamically defining an instance variable of the same name in the current `ModuleHook` instance (e.g., `self.datas`). These variables would then be publicly accessible in the standard way without having to go through the `__getattr__()` magic lookup process. Is that right? If so, I begrudgingly admit: **that works, too.**

It's also simpler and more efficient. So, I'm sold. We'll go with that. I kind of liked firewalling magic attributes from external access into a private `dict`, but we can always privatize this later if we really need to. (_We probably won't._)